### ISSUE TITLE
BPE-drop question.

### ISSUE THREAD

**Issue by TomtheCodeBot** [neelsjain/baseline-defenses#2]
Good day to the devs.

Thank you so much for your work, it establishes the baseline methods to defend LLM against adversarial attacks.

I am currently replicating the results for the BPE-drop. In this process, did you include the system prompt in the bpe-drop or not? If you didn't, could you share with us the piece of code you used to tokenize the prompt separately without interfering with the system prompt?

Thank you for your time.

**Comment by neelsjain** [comment]
Hello, 

Thank you for your message.
We did not apply the BPE dropout to the system prompt. We individually tokenized the system prompt and instruction. How you do the BPE dropout depends on the code base preprocessing script.

Here is an example, but I expect your's might be different:

```
def break_the_tokens(batch, tokenizer, bt_alpha, conv_template):
    
    break_strings = conv_template.roles
    sep = conv_template.sep
    if any(name in tokenizer.name_or_path for name in ["alapaca", "Alpaca"]):
        pad_token_id = tokenizer.pad_token_id
        bos_token_l = [tokenizer.bos_token_id]
        break_format = f'{break_strings[0]}:|{sep}{break_strings[1]}:'
    elif any(name in tokenizer.name_or_path for name in ["vicuna","llama"]):
        pad_token_id = tokenizer.pad_token_id
        bos_token_l = [tokenizer.bos_token_id]
        break_format = f'{break_strings[0]}:{sep}|{sep}{break_strings[1]}:'
    elif any(name in tokenizer.name_or_path for name in ["guanaco","chatglm2"]):
        sep2 = " "
        pad_token_id = tokenizer.pad_token_id
        bos_token_l = [tokenizer.bos_token_id]
        break_format = f'{break_strings[0]}:{sep2}|{sep}{break_strings[1]}:'
    elif any(name in tokenizer.name_or_path for name in ["falcon"]):
        sep2 = " "
        pad_token_id = tokenizer.encode(tokenizer.special_tokens_map['pad_token'])[0]
        bos_token_l = []
        break_format = f'{break_strings[0]}:|{sep}{break_strings[1]}:'
    
    else:
        raise Exception("Case not written")
    explore = False
    break_sep = conv_template.sep
    b0, b1, b2 = [], [] , []

    for text in batch: 
        result = re.split(break_format, text)
        b0.append(result[0]+f'{break_strings[0]}:')
        b1.append(result[1])
        b2.append(f'{sep}{break_strings[1]}:' + result[2])
    
    sep_input_ids = []
    for i, b in enumerate([b0,b1,b2]):
        if i ==1:
            if any(name in tokenizer.name_or_path for name in ["vicuna","llama","alpaca" ,"guanaco"]):
                if bt_alpha == 0:
                    op = tokenizer.sp_model.encode(b, alpha=bt_alpha, enable_sampling=False)
                else:
                    op = tokenizer.sp_model.encode(b, alpha=bt_alpha, enable_sampling=True)
            elif any(name in tokenizer.name_or_path for name in ["falcon"]):
                tokenizer._tokenizer.model.dropout = bt_alpha
                op = tokenizer(b).input_ids
            else:
                raise Exception("case not written")
        else:
            if any(name in tokenizer.name_or_path for name in ["vicuna","llama","alpaca","guanaco"]):
                op = tokenizer.sp_model.encode(b, enable_sampling=False)
            elif any(name in tokenizer.name_or_path for name in ["falcon"]):
                tokenizer._tokenizer.model.dropout = 0.0
                op = tokenizer(b).input_ids
        sep_input_ids.append(op)

    input_ids = list(bos_token_l + list(itertools.chain.from_iterable(list(t))) for t in zip(sep_input_ids[0], sep_input_ids[1],sep_input_ids[2]))

    padded_input_ids = torch.nn.utils.rnn.pad_sequence([torch.tensor(i[::-1],dtype=torch.int) for i in input_ids],  batch_first=True,padding_value= pad_token_id).flip(dims=[1]) 
    
    attentions = torch.ones_like(padded_input_ids).masked_fill(
            padded_input_ids == tokenizer.pad_token_id, 0
        )

    
    batch_ips = {'input_ids':padded_input_ids, 'attention_mask': attentions}
    return batch_ips
```

Let me know if you have more questions.

**Comment by TomtheCodeBot** [comment]
Got it, thank you for your help. I think it would be useful if this is in the repository as well.

Once again, thank you for your time and help.
